### ISSUE TITLE
Address Connection Drop In Manual

### ISSUE THREAD

**Issue by drag05** [rstudio/pool#129]
Hello,
I am writing millions of rows to a `MySQL` database using a `Rcpp` iterator. At the end of each iteration, the application appends the rows resulted from that iteration to a database table on the server. The server is on a cloud.
There could be thousands of rows/iteration. The process is lengthy making connection drop likely.
Eventually, the connection drop is detected at the end of iteration when the writing to database is ready to start. 
When the connection is found dead, from the error messages I see that `dbPool` swaps objects in the pool trying to reconnect (or so I think, as the documentation lacks clarity and specifics at times).

Therefore I hope you could help me with these questions:

-  what are the "checked out" objects and why would they be needed? Could there be any examples showing their advantage over just using `conn = dbPool(...)`? 
-  does the argument `validationInterval` apply to objects in pool or to current connection or, only to checked out objects?
-  is the pool diminished by the `idleTimeout` and why?
-  is `dbPool` set to reconnect automatically if the connection is found dead by switching to the next object in the pool?
- are all connection objects in the pool dead when the connection drop originates at the server end? And if so,
-  would it help inserting a `dbIsValid` check followed by a custom function to reconnect in case 
`dbIsValid` is found **FALSE**?
- in case of successful reconnection will the writing continue from where it was interrupted (the iteration result resides in memory during this time) or will it start all over?
I could be wrong but I have found no mention on how connection drop is treated in `pool`'s Manual. 
Please try to address and/or clarify the procedures for these situations.
Thank you!

**Comment by hadley** [comment]
If you're writing millions of rows all at once, you probably don't want to use pool.

**Comment by drag05** [comment]
@hadley  Why?

**Comment by hadley** [comment]
Because you presumably want to optimise for performance, which means getting as close to the database as possible.

**Comment by hadley** [comment]
The point of the pool package is to manage very long running connections, as you might need in a shiny app. If you're creating millions of rows, you're probably doing it all at once, so I don't see pool being that useful. But perhaps you could explain your use case in more detail?

**Comment by drag05** [comment]
@hadley To make a long story short, I was sending to a cloud database subsets of data frames resulted from permuting with repetition and with replacement a set of objects to a set of locations along with some other calculations. 

For this, I have used the Iterator from `RcppAlgos` package which has the advantage of allowing control of the number of iterations/cycle i.e. the number of rows to be sent to the database after one cycle of iterations is complete.  Also, `data.table` package set for 12 logical cores, with 4 to spare.

The entire exercise yielded close to 1 billion rows sent sequentially to 2 database tables. One of the tables was small, around 70 million rows the other however took the bulk of the data.
Aware of this large amount, I tried initially to send between 10^5 and 50^5 rows per cycle. 

Realizing that connection stayed idle too long while the iterator was crunching (increasing the value of `idleTimeout` argument didn't seem to help much!), I settled with sending 10^4 rows/cycle. 

The idle time and prevention of connection drop were the reasons I opted for `pool` instead of `DBI`. 

I now realize one possible mistake I made then, namely connecting to the database too early, at the moment the iterator started a new cycle instead of connecting at the end the cycle when the rows were ready to be sent to the database. That could have probably saved a few days, maybe a week.

Other factors affecting the connection could have been the fact that the database was not local. Toward the end of the exercise the drop frequency increased forcing me to write the last few hundred million rows to disk, to a `.csv` file which then was sent to the database. This operation was a bit faster as it only involved reading-then-writing.

Any way, the task was completed long ago. Any insight would be appreciated, Thank you!

**Comment by jcheng5** [comment]
Connection pools are primarily useful in my experience where you are 1) creating connections quite often, 2) performing short-running queries, and 3) trying to optimize for low latency. If all three are true, then the complexity of a connection pool is worth it for the performance impact of not having to create new connections—an operation that might take a couple of hundred milliseconds.

In your case, you’re doing very large-grained transactions against a database that seems to have an aggressive timeout (relative to the frequency of your using the connection, at least); there’s no benefit to maintaining a connection pool since the overhead of opening the connection is trivial compared to what you’re doing with it.

100% agreed with you that you should wait to open the connection until you are actually going to use it.

As far as connection drops in pool, it sounds to me like you were calling poolCheckout at the beginning of your iteration? While the connection is checked out, pool cannot help you recover from that connection being dropped. Instead, pool is supposed to detect dropped connections at the time of poolCheckout (or just using the pool object as a DBI connection object) and return a fresh one to you if necessary.

**Comment by drag05** [comment]
@jcheng5 I have misinterpreted the role of the connection pool. I was thinking that the pool is there to replace any dropped connection should this happen. It would be very helpful adding more insights to the `pool` Manual. 
Also, I think a connection drop automatic recovery feature would be helpful with any `R` database package. It seems `ruby` has such [feature](https://gist.github.com/shinzui/250080). There is also a `python` [package](https://gist.github.com/shinzui/250080) for `Ubuntu`. Maybe it is the time `R` had one too. 
I have tried using an error catching function - found somewhere on `stack exchange` - which did not work as expected maybe because the reason for drop was at either end of the connection. The database I was mentioning was a MySQL database which seems to have a connection timeout [feature](https://sebhastian.com/lost-connection-mysql-server-during-query/) built-in. This is probably helpful when the drop happens at the server end.
Thank you!
